### PR TITLE
[FIX] Rewrite inheritance principle

### PR DESCRIPTION
**Pull request by Lestropie**
Draft PR in search of feedback.

This change relates to #102, #259, #482; am escalating due to re-invigoration of BEP016, in particular bids-standard/bids-bep016#32.

IMO, there are multiple issues with the current description of the inheritance principle:

1. Some details are reiterated in different ways, which shouldn't be necessary if the principle is stated unambiguously;
2. Despite the availability & use of filesystem tree examples, many aspects of the principle are described through *hypothetical* modifications rather than presenting such;
3. Some instances of point 2 result in reference to either files that don't exist or would need to be renamed in order to satisfy the change suggested;
4. There is a persistent blurring between official description of the principle that should be interpreted literally, and recommendation / guidance.
5. Inheritance is described with respect to the lower-level JSON file, rather than with respect to a data file to which multiple JSONs may be applicable.
6. The logic by which software tools should classify which JSONs are or are not applicable to a given data file is not adequately clear.

Given the breadth of these issues I think a re-write is warranted. I am completely open to proposals for modifications or proposal of outright alternatives, but the ball needs to start rolling somewhere.

Full disclosure: as per #259, I would like to pursue the prospect of modifying the inheritance principle itself, not just the description of such; specifically removal of the preclusion of having multiple applicable JSONs at one level of the hierarchy. However here I am focusing solely on modifying the text describing the principle as it currently stands, which would be applicable to a patch update. Changing the principle itself should IMO be deferred to a minor or potentially major update; in addition, the isolation of that specific aspect and modification thereof should be somewhat facilitated by the clarification of the current state of the principle following these changes.

### CODE REVIEW

**Comment by Remi-Gau** [comment]
pinging @yarikoptic and @VisLab as they both have expressed interest in this issue in the past

File: src/02-common-principles.md
Comment (Remi-Gau) [inline src/02-common-principles.md:412]:
```suggestion
Any metadata file (such as `.json`, `.bvec` or `.tsv`) MAY be defined at any directory level. 
For a given data file, any metadata file at that directory level or higher 
is applicable to that data file if:
- the metadata file and the data file possess the same suffix,
- the metadata file does not include any entities absent from the name of the data file. 
Such files are loaded from the top of the directory hierarchy downwards, 
such that values from the top level are inherited by all data files 
at lower levels to which it is applicable unless overridden 
by a value for the same key present in another metadata file at a lower level 
(though it is RECOMMENDED to minimise the extent of such overrides). 
There is no notion of "unsetting" a key/value pair.
```

Comment (Lestropie) [reply src/02-common-principles.md:412]: Yes, I'd contemplated bullet-pointing this myself; or potentially even enumerating so that datasets violating specific requirements / software code can be cross-referenced more precisely?

@VisLab: Good catch. There's going to need to be a hard split between JSON and other metadata files. For `.bvec` / `.bval` / `.tsv`, it I think needs to be simply selection of whichever applicable file is lowest in the directory tree; anything above it gets ignored, there's no reason to describe the contents of such as being "overwritten" by the content of files lower down. Only edge case that comes to mind would be `.tsv` files that happen to contain exactly the same columns (including their titles), but even then it's hard to justify choosing between a concatenation and a replacement. Are there any other metadata types that are worth being familiar with for the sake of getting the language / logic of the inheritance principle right?

Comment (Remi-Gau) [reply src/02-common-principles.md:412]: Good idea about using an ordered list and referring to those points in the example.

Comment (VisLab) [reply src/02-common-principles.md:412]: So am I understanding that for JSON the resolution is top-down, with top (closest to the root) having precedence.
For the others it is bottom up, with the bottom (farthest from the root) having precedence?  

The notions of "lowest" and "highest" level in the tree isn't necessarily clear.  Does lowest mean farthest from the root? 

I honestly thought only JSON files got inherited in the sense that there could be multiple applicable ones for the same suffix.

Comment (Remi-Gau) [reply src/02-common-principles.md:412]: I think the only case of "official" of TSV inheritance we have around is having a single `task-foo_events.tsv` in the root of the directory in case all subjects / runs had the same design:

see from the BIDS examples

```bash
╰─⠠⠵ ls -l */*task*tsv
-rw-rw-r-- 1 remi remi  143 Nov 10 11:33 ds114/task-covertverbgeneration_events.tsv
-rw-rw-r-- 1 remi remi  280 Nov 10 11:33 ds114/task-fingerfootlips_events.tsv
-rw-rw-r-- 1 remi remi  143 Nov 10 11:33 ds114/task-overtverbgeneration_events.tsv
-rw-rw-r-- 1 remi remi  127 Nov 10 11:33 ds114/task-overtwordrepetition_events.tsv
-rw-rw-r-- 1 remi remi 1054 Nov 10 11:33 eeg_ds000117/task-facerecognition_channels.tsv
-rw-rw-r-- 1 remi remi 1807 Nov 10 08:42 synthetic/task-nback_events.tsv
```

Comment (VisLab) [reply src/02-common-principles.md:412]: We probably need to expand the discussion of events inheritance.  Inherited `events.tsv files` are needed at several levels.  

For example, because BIDS forces multi-modality data to be separated into separate directories, simultaneously recorded MEG/EEG or fMRI/EEG will have the same event files for each run.  

In the following example from [hed-examples](https://github.com/hed-standard/hed-examples/tree/main/datasets/eeg_ds003654s_inheritance), EEG and MEG were recorded simultaneously, but separated into separate modality directories.  

```
ds_003654s:
       task-FacePerception_events.json
       sub-02:
               sub-002_task-FacePerception_events.json
               sub-002_task-FacePerception_run-1_events.tsv
               eeg:
                      sub-002_task-FacePerception_run-1_eeg.set 
               meg:
                      sub-002_task-FacePerception_run-1_meg.fif
   . . .
```

Right now we are assuming that multiple `events.json` files can be applicable to a given `events.tsv` files and that the keys are resolved from directory root down in the JSON files.  I think this is the right behavior for JSON files, but it needs to be clarified in the specification.

We have only made a single `events.tsv`  applicable to any given recording, but that `events.tsv` file could appear anywhere in the directory hierarchy.   

Events are a special case, and I think it might be feasible/useful to allow multiple events files at different levels to apply to a given run. The events are the union of the rows/join of columns.  This might be too complicated.  

The issue also comes up with derivatives.  Here the derivatives might have their own set of events. If the events from the raw data also apply do they need to always be copied into the events files? There was some discussion of this in the past, but there didn't seem to be a clear resolution or consensus on this when the discussion occurred.

Comment (VisLab) [reply src/02-common-principles.md:412]: I also have a general question about inheritance.  Is the matching of entity components and their names case sensitive?

Comment (Lestropie) [reply src/02-common-principles.md:412]: > The notions of "lowest" and "highest" level in the tree isn't necessarily clear. Does lowest mean farthest from the root?

I always have to cross-reference elsewhere in the spec to remind myself which is which. Does anyone know if there's a robust reference-able definition somewhere? Or does this need to be defined upfront in the spec?

**Comment by Lestropie** [comment]
Lots of changes (not expecting this to be a quick PR). I think that it is by necessity transitioning from the original text being along the lines of "here's how you should do this thing" to what is really required here for the sake of stringent enforcement and validation being "here's the rules", with the downstream *consequences* of such expressed in a more user-friendly way separate to the rules themselves (what's currently called "corollaries" here could conceivably be expanded quite a lot).

Note that while @Remi-Gau's change proposal above was merged, I've explicitly unresolved the conversation as there is a lot of content there unrelated to that specific proposal that have not been resolved. Would suggest branching separate chains of thought into separate comment threads tied to the relevant lines; this one is invariably going to be messy, best to mitigate it where we can.

**Comment by VisLab** [comment]
I have reviewed the commits and the bullet points make things much clearer.  Thank you!

File: src/02-common-principles.md
Comment (yarikoptic) [inline src/02-common-principles.md:552]:
```suggestion
1.  A metadata file MUST NOT have a filename that would be otherwise applicable
```
if there would be agreement on https://github.com/bids-standard/bids-specification/pull/962 prompted by this line ;-)

File: src/02-common-principles.md
Comment (yarikoptic) [inline src/02-common-principles.md:557]: FWIW: it is such a nice concise rule which helps to avoid ambiguity and the workaround in Example 3 is quite cute ;)

File: src/02-common-principles.md
Comment (yarikoptic) [inline src/02-common-principles.md:547]: although markdown does numbering for us, since there are direct references to them in the text (e.g. `(rule 5.2.2)`) I think it would be better just to number them here as well, so it is easier to refer back while reading markdown and not rendered version

**Review by yarikoptic — APPROVED** [review]
I really like it, well done.  IMHO should be taken out of the draft! ;)

Left some minor comments

File: src/02-common-principles.md
Comment (yarikoptic) [reply src/02-common-principles.md:557]: OH -- I remembered what bothered me in my original #102 and outlined in Edit 1 there. Citing an example from there:

> I placed myself into a corner with an example of having e.g.
> 
> ```
> sub-1_task-task1_run-1_bold.json and
> sub-1_task-task1_acq-X_run-1_bold.json
> ```
>
> per subject (should be ok), and then trying to aggregate over them while retaining also _acq- if defined. 

and there are many other entities/use-cases which would fall into similar situation.  What I am thinking is to add a clarification as subitem here or a separate rule:

- A metadata file at the data files level of hierarchy MUST not be considered for inheritance if there is a matching in entities data file.

This would make `sub-1_task-task1_run-1_bold.json` not a to be considered for providing metadata to enrich `sub-1_task-task1_acq-X_run-1_bold.json`, and examples below would stay valid (we might want to add an example for this rule though)

Comment (Lestropie) [reply src/02-common-principles.md:557]: > it is such a nice concise rule which helps to avoid ambiguity and the workaround in Example 3 is quite cute ;)

Well, actually (from first post):

> Full disclosure: as per #259, I would like to pursue the prospect of modifying the inheritance principle itself, not just the description of such; specifically removal of the preclusion of having multiple applicable JSONs at one level of the hierarchy.

😬 

The proposed changes here do at least I think demonstrate what would need to change in order to facilitate such. Any text I propose for such will need to be very clear for both users and software creators exactly what is permitted vs. not permitted and in what order JSONs should be loaded. But as I said in the original post, this is not a change in descriptive language but a change in prescriptive permissible structure, and so IMO necessitates either a minor or major version change rather than just a patch.

Comment (yarikoptic) [reply src/02-common-principles.md:557]: Gotcha. Given the currently loose formulation of the principle, depending on the changes to it, I think it might be feasible to get it done within minor revision. So it would be great to see this PR be finalized/merged soon.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md:547]: Better would be enumeration cross-references; but as far as I can tell no standardized solution for such exists?

Comment (Lestropie) [reply src/02-common-principles.md:547]: > it would be better just to number them here as well

Not permitted: Lestropie/bids-specification#2

Comment (yarikoptic) [reply src/02-common-principles.md:547]: eh, thanks for trying it out! 
Indeed from https://www.npmjs.com/package/remark-lint-ordered-list-marker-value I see that
- the default is `"one"`
- there seems to be no I do not see some way to override locally in a file/for this specific list.
   - https://stackoverflow.com/a/63753061/1265472 suggests that a local to a directory `.remarkrc` could be used to overload some settings, but that would not help here/more confusing than helpful in the long run
- the other places in BIDS specification follow the "one" convention and there it seems to make sense

so I guess I will "retract" my invaluable comment ;-) Thanks again for trying

Comment (yarikoptic) [reply src/02-common-principles.md:547]: actually according to https://github.com/remarkjs/remark-lint#ignore-warnings there is an inline way to disable/re-enable specific warnings in a specific block, but I do not see there a way to change configuration setting (i.e. make it "ordered") . I will suggest this disable/reenable in the PR just to see if works -- not sure if worth it if we can't lint the order

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:554]: I like these back-references to 2.2, 2.3 etc. -> however in the rendered version, the nested enumerated list has letters, not numbers (see screenshot below)

image

So I would suggest to refer to 2a, 2b, 2c here.

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:562]:
```suggestion
        (for instance, [`bvec` / `bval` files for diffusion MRI](#04-modality-specific-files/01-magnetic-resonance-imaging#required-gradient-orientation-information)),
```

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:563]: could we use MUST instead of SHALL here? Simply because so far there is zero mention of SHALL in the spec, but lots of MUST - and I thought these were equivalent?

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:603]:
```suggestion
Contents of file `task-rest_bold.json`:
```

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:612]:
```suggestion
Contents of file `sub-01/func/sub-01_task-rest_acq-longtr_bold.json`:
```

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:623]: Rule 2.c, see my comment above.

I see that below this would be "5.b.ii" instead of 5.2.2, so maybe we need to discuss my suggestion - as 5.b.ii is kind of ugly.

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:627]:
```suggestion
the value for field "`RepetitionTime`" is therefore overridden to the value `3000.0`.
```

**Review by sappelhoff — COMMENTED** [review]
This is wonderful work, thanks a lot! I left a few comments

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md:563]: I had thought intuitively that there was a distinction between MUST and SHALL, but that appears to be incorrect based on RFC2119. My thinking here was that it's entirely plausible that an app may have no interest in the content of a sidecar file, and just load an e.g. image in isolation, and therefore stating that such sidecar information MUST be loaded is to me unnecessarily dictatorial. It's more like: if you are loading a data file, *and* you are interested in the sidecar data associated with that file, *then* only the lowest sidecar file MUST be utilised.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md]: Closed by 8e808a12.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md]: Closed by 8e808a12.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md]: See if you're happy with 8e808a12.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md:603]: Closed by 8e808a12.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md:612]: Closed by 8e808a12.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md]: Revised by 8e808a12, but not yet resolving in case the enumeration formatting needs to be discussed further.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md:627]: Closed by 8e808a12.

**Comment by Lestropie** [comment]
Invested parties please also see Lestropie/bids-specification#3, which proposes an alteration to the contents of this PR.

File: src/02-common-principles.md
Comment (sappelhoff) [reply src/02-common-principles.md]: yes, thank you - resolving.

File: src/02-common-principles.md
Comment (sappelhoff) [reply src/02-common-principles.md]: Looking at it again, I don't find it ugly anymore :-) Thanks for making the change.

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:572]: nitpick: Are we using American or British spelling throughout the spec? I had assumed we use American spelling, and indeed there are two occurrences of `minimize` and none of  `minimise` in the spec currently. Perhaps we should add a CI step (in a different issue/PR) that checks for American/British spelling. Just to be consistent.

Comment (Remi-Gau) [reply src/02-common-principles.md:572]: Not a bad idea, but I am not quite ready to write that PR. :-)

File: src/02-common-principles.md
Comment (Remi-Gau) [reply src/02-common-principles.md]: > Is the matching of entity components and their names case sensitive?

I think the case sensitivity issues was solved in https://github.com/bids-standard/bids-specification/pull/858

**Comment by Remi-Gau** [comment]
Generally in favor of merging this after https://github.com/Lestropie/bids-specification/pull/3 is in.

**Review by sappelhoff — APPROVED** [review]

**Comment by Lestropie** [comment]
@sappelhoff I'm done tinkering, content as it currently stands. Will have a go at #259 in a separate PR as discussed.

"`linkchecker`" CI test failed, naively it looks like it might have been a one-off failure that just requires a re-run since it doesn't seem to me to relate to the proposed changes?

**Comment by sappelhoff** [comment]
> "linkchecker" CI test failed, naively it looks like it might have been a one-off failure that just requires a re-run since it doesn't seem to me to relate to the proposed changes?

That issue was solved in `master`, so I merged those changes back here - should all be fine now.

>  I'm done tinkering, content as it currently stands. 

awesome, let's try to get a final approval by a bunch of people and merge this!

**Review by Remi-Gau — APPROVED** [review]

**Comment by VisLab** [comment]
There is a problem with the bullet point numbering when I do the view file on GitHub.  The first set of bullet points uses 2 (with subcategories i, ii, and iii), but the text refers to 2 (a, b, c).  

Similarly in the corollaries, rule three refers to "per rule 5.2"

**Review by effigies — COMMENTED** [review]
Overall looks good. Two minor issues.

File: src/02-common-principles.md
Comment (VisLab) [reply src/02-common-principles.md:583]: I'm okay with taking out the recommended here.

**Review by VisLab — APPROVED** [review]

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md]: Closed by 33cc25a.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md]: Closed by 33cc25a.

File: src/02-common-principles.md
Comment (Lestropie) [reply src/02-common-principles.md:583]: >  This is a best practice written in the language of a validatable rule.

1.  This seems to extend beyond the RFC2119 definition. Is this an established BIDS-specific interpretation that is documented somewhere?

2.  Would the same criticism not also apply to 5.2?

    > For [JSON files](#key-value-files-dictionaries), key-values are loaded from files from the top of the directory hierarchy downwards, such that key-values from the top level are inherited by all data files at lower levels to which it is applicable unless overridden by a value for the same key present in another metadata file at a lower level (though it is RECOMMENDED to minimise the extent of such overrides).

    Issuing a validator warning regarding the presence of individual key-value overrides may be unexpected.

**Comment by Lestropie** [comment]
> There is a problem with the bullet point numbering when I do the view file on GitHub.

I think the most important correspondence here is going to be not with a GitHub rendering, but with the website / PDF renderings. I couldn't find an existing example; perhaps if someone could duplicate this branch on the main repo, then set up ReadTheDocs to render that branch, we could check what the current behaviour is there?

**Comment by sappelhoff** [comment]
> perhaps if someone could duplicate this branch on the main repo, then set up ReadTheDocs to render that branch, we could check what the current behaviour is there?

We should have all the needed checks available from the CI:

- [pdf](https://10077-150465237-gh.circle-artifacts.com/0/bids-spec.pdf)
- [html](https://bids-specification--946.org.readthedocs.build/en/946/)

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:586]:
```suggestion
1.  Where multiple applicable JSON files are loaded as per rule 5.b, key-values can
```

File: src/02-common-principles.md
Comment (sappelhoff) [reply src/02-common-principles.md:583]: >  Is this an established BIDS-specific interpretation that is documented somewhere?

AFAIK this is not documented anywhere, but among maintainers we have often discussed that RECOMMENDED corresponds to a "warning" level in the validator, and MUST corresponds to an "error" level in the validator. The main point of this consideration is to not design a specification that we cannot reasonably validate.

Having that said, there are many cases in the spec where we RECOMMEND or require (MUST), where validation is currently not happening and might be difficult to implement --- as you correctly point out in your second point.

Personally I am fine with having some recommendations that we cannot "warn" about (see especially 5.b here, but maybe also the point raised by Chris.).

File: src/02-common-principles.md
Comment (sappelhoff) [inline src/02-common-principles.md:572]:
```suggestion
        (though it is RECOMMENDED to minimize the extent of such overrides).
```

**Comment by sappelhoff** [comment]
Thanks for leading this effort @Lestropie! I think it's a huge improvement in terms of clarity.

**Pull request merged**
